# 12 · Crop maintenance: spend labor where it changes production

**Kaggriculture | controlled feature-family experiment**

**Decision:** do not promote notebook 11's route intervention. Its primary pair lost 2 own coins, gained 1 coin of margin, and remained a win. The registered conservative gate stopped it.

**Question:** can a next-refresh WATER eligibility feature free useful worker actions without sacrificing survival, production, or final cash?

This notebook changes **one job-eligibility predicate** on zero-indexed days **20 and 21**. It does not change priorities, layout, routing, hiring, purchasing, or the final-sale rule. Deferring maintenance transfers work to tomorrow; it is not automatically safe.

**Work limits:** 120-second screen; a separate 180-second paired pilot only after activation; 500 ms per callback. No installations, cloud-resource changes, Git writes, or automatic scale-up. Stop on an error and use the bundle instructions at the end.

In [1]:
from pathlib import Path
import json, os, signal, subprocess, sys, time
import pandas as pd
import plotly.io as pio
from IPython.display import display

ROOT = Path.cwd().resolve()
if not (ROOT / "run_maintenance.py").exists():
    candidate = Path.home() / "kaggriculture_water_maintenance"
    if not (candidate / "run_maintenance.py").exists():
        raise FileNotFoundError("Open this notebook from the extracted package folder.")
    ROOT = candidate
sys.path.insert(0, str(ROOT))
from visualize_maintenance import prior_figure, screen_figures, pilot_figures, dashboard
OUT = ROOT / "outputs"
pio.renderers.default = "plotly_mimetype"
figures = []
print("LIVE MANUAL WORKSPACE — no prior-stage reruns")
print("Package:", ROOT)
print("Interpreter:", sys.executable)

LIVE MANUAL WORKSPACE — no prior-stage reruns
Package: /home/sagemaker-user/kaggriculture_water_maintenance
Interpreter: /home/sagemaker-user/projects/kaggriculture/.venv/bin/python


## 1 · Preserve the result that stopped the previous direction

The screen and endpoint measures answer different questions. Static assignment utility improved in 42 of 69 sampled states, but the tested continuous trajectory did not increase own terminal cash. The +1 margin is real and must not be described as negative; it simply does not satisfy the preregistered conservative gate.

The following table is copied from your returned notebook-11 result, not a result from this notebook.

In [2]:
review = json.loads((ROOT / "reference/input_review.json").read_text())
previous = pd.read_csv(ROOT / "reference/notebook11_pairs.csv")
display(pd.Series(review, name="verified prior evidence").to_frame())
fields = [c for c in ["role", "seed", "seat", "arm", "coins_delta", "opponent_coins_delta", "coin_margin_delta", "match_score_delta"] if c in previous]
display(previous[fields])

,verified prior evidence
bundle_files_verified,28
archive_sha256,4b141bc88ade1f920cfaaf34eb7740d133902ccf123d07...
screen_report_sha256,9f290b0cd6cabfb198c895daaf0d27476283b189d782be...
pilot_report_sha256,404ebd66312d5a6db9e16e335e2fa2b85101826ce2566f...
executed_code_cells,14
notebook_error_outputs,0
prior_decision,STOP_NEGATIVE_ENDPOINT
primary_own_cash_delta,-2
primary_opponent_cash_delta,-3
primary_coin_margin_delta,1


,role,seed,seat,arm,coins_delta,opponent_coins_delta,coin_margin_delta
0,negative_control,1601,0,sequential,0.0,0.0,0.0
1,primary,1601,0,coordinated,-2.0,-3.0,1.0


In [3]:
fig = prior_figure(ROOT / "reference")
figures.append(fig)
fig.show()

## 2 · Register the feature intervention before reading its outcomes

Thirty per-plant descriptors represent immediate WATER yield, next-refresh survival and units, crop cadence, fertilizer state, decay, and next-day maintenance debt. Eleven state summaries represent task pressure and available workers. Several fields are existing primitives or context; they are not all newly discovered signals.

The two next-refresh scenarios hold all intervening treatments, harvesting, and replanting fixed and include the remaining in-day decay phases. Neither scenario uses rewards, a seed, hidden opponent inventory, or future shops.

The candidate retains WATER for a dry-streak survival obligation, immediate units, or a next-refresh survival/yield gain. Otherwise it may defer it during the registered two-day window. The original ranker chooses among all remaining jobs. A null fork must exactly reproduce the original rule on the same observations.

In [4]:
protocol = json.loads((ROOT / "PROTOCOL.json").read_text())
display(pd.Series(protocol, name="registered design").to_frame())
dictionary = pd.read_csv(ROOT / "feature_dictionary.csv")
display(dictionary.groupby("table").size().rename("documented fields").to_frame())
display(dictionary.loc[dictionary.feature.isin(["water_bonus_now", "next_units_delta", "survival_gain", "next_day_maintenance_debt", "defer_eligible", "keep_water"])])

,registered design
experiment,water-maintenance-deferral-v1
source_commit,7194116dfc92a8663139611233b5a221dad431a4
active_days_zero_indexed,"[20, 21]"
eligible_observation_days_zero_indexed,"[0, 28]"
screen_sampling,"hours 0,6,12,18 each day 0..28 plus all hours ..."
screen_states,468
screen_source_keys,"[[1601, 0], [1601, 1], [1602, 0]]"
selection,First source by seed/seat with any within-wind...
screen_cap_seconds,120
pilot_cap_seconds,180


,documented fields
table,
plant,30
state,11


,table,feature,units,definition,availability,adoption_status
9,plant,water_bonus_now,product units,Capped immediate change in yield from WATER no...,"Current own public farm, visible prices, publi...",Candidate/context; not independently validated...
15,plant,next_units_delta,product units,WATER-now minus no-WATER next-refresh yield; b...,"Current own public farm, visible prices, publi...",Candidate/context; not independently validated...
16,plant,survival_gain,indicator difference,WATER-now survival minus no-WATER survival at ...,"Current own public farm, visible prices, publi...",Candidate/context; not independently validated...
19,plant,next_day_maintenance_debt,indicator,Eligible deferral leaves a surviving plant wit...,"Current own public farm, visible prices, publi...",Candidate/context; not independently validated...
20,plant,defer_eligible,indicator,"Unwatered, refresh available, no dry-streak ob...","Current own public farm, visible prices, publi...",Candidate/context; not independently validated...
21,plant,keep_water,indicator,Keep the source WATER job if unwatered and sur...,"Current own public farm, visible prices, publi...",Candidate/context; not independently validated...


## 3 · Run the bounded saved-observation screen

This stage verifies package/dependency fingerprints, the reviewed checkout, the pinned engine, and all ten restored objects. It does not rerun earlier notebooks.

Before any research observation is processed, **736 artificial crop fixtures** compare the standalone scenario calculation against the installed official WATER, decay, and nightly-refresh primitives, in both WATER/no-WATER paths. These are correctness fixtures, not games or performance results.

The screen samples 468 states from the three preserved coordinated episodes, including every hour in the intervention window. Only changed actions inside the registered window can activate the pilot. Source selection is the first activated seed/seat, never a choice based on rewards.

The stage prints UTC heartbeats, saves each completed source, and rejects an unchanged failed or interrupted restart.

In [5]:
receipt = Path.home() / "kaggriculture_manual_resume/state/runtime.json"
if not receipt.exists():
    raise FileNotFoundError("Existing runtime receipt is missing. Do not reinstall; bundle the error.")
PYTHON_BIN = json.loads(receipt.read_text())["executable"]
if Path(sys.executable).absolute() != Path(PYTHON_BIN).absolute():
    raise RuntimeError("Select Kernel → Kaggriculture Manual (verified source), then restart this notebook.")

# Streams child logs without blocking the notebook's emergency deadline.
# The runner owns its inner worker's hard cap and termination.
def run_stage(stage):
    caps = {"screen": 120, "pilot": 180}
    if stage not in caps:
        raise ValueError("Unregistered stage")
    OUT.mkdir(exist_ok=True)
    logpath = OUT / (stage + "_notebook_console.txt")
    process = None
    offset = 0
    started = time.monotonic()
    def show_new():
        nonlocal offset
        with logpath.open("r", errors="replace") as stream:
            stream.seek(offset)
            text = stream.read()
            offset = stream.tell()
        if text:
            print(text, end="", flush=True)
    try:
        with logpath.open("w") as console:
            process = subprocess.Popen(
                [PYTHON_BIN, "-u", str(ROOT / "run_maintenance.py"), stage],
                cwd=ROOT, stdout=console, stderr=subprocess.STDOUT,
                start_new_session=True,
            )
            while process.poll() is None:
                show_new()
                if time.monotonic() - started > caps[stage] + 30:
                    raise TimeoutError("Notebook emergency deadline; bundle diagnostics.")
                time.sleep(0.25)
            show_new()
            if process.returncode:
                raise RuntimeError(f"{stage} failed. Stop; save and bundle. See {logpath}.")
    except BaseException:
        if process is not None and process.poll() is None:
            os.killpg(process.pid, signal.SIGTERM)
            try:
                process.wait(timeout=5)
            except subprocess.TimeoutExpired:
                os.killpg(process.pid, signal.SIGKILL)
                process.wait()
        raise
    return json.loads((OUT / stage / "report.json").read_text())

In [6]:
screen = run_stage("screen")
display(pd.Series({k:v for k,v in screen.items() if k not in ("output_sha256",)}, name="live screen result").to_frame())
mechanics = json.loads((OUT / "mechanics.json").read_text())
print("Installed-engine component checks:", mechanics["status"], mechanics["fixtures"], "fixtures")

...............................................................
----------------------------------------------------------------------
Ran 63 tests in 4.067s

OK
{"utc": "2026-09-12T02:34:46.973700+00:00", "stage": "PREFLIGHT_PASSED", "verified_objects": 10}
{"utc": "2026-09-12T02:34:47.268223+00:00", "stage": "MECHANICS_PASSED", "fixtures": 736}
{"utc": "2026-09-12T02:34:51.142146+00:00", "stage": "HEARTBEAT", "name": "screen", "elapsed_seconds": 10.0}
{"utc": "2026-09-12T02:34:52.091897+00:00", "stage": "SCREEN_EPISODE_SAVED", "seed": 1601, "seat": 0, "opponent": "livestock_fertilizer", "arm": "coordinated", "states": 156}
{"utc": "2026-09-12T02:34:55.654088+00:00", "stage": "SCREEN_EPISODE_SAVED", "seed": 1601, "seat": 1, "opponent": "livestock_fertilizer", "arm": "coordinated", "states": 156}
{"utc": "2026-09-12T02:34:59.621648+00:00", "stage": "SCREEN_EPISODE_SAVED", "seed": 1602, "seat": 0, "opponent": "livestock_fertilizer", "arm": "coordinated", "states": 156}
{"utc": "2026-09-

,live screen result
action_changes,51
candidate_callback_max_ms,30.470355
cloud_resources_modified,False
code_sha256,9c952471dc3d6a9185d5bda803955b323ca3c0cf0de426...
crop_rows,3658
decision,ELIGIBLE_FOR_ONE_PAIRED_PILOT
elapsed_seconds,18.623058
engine_sha256,bc8a54879ef02c7ea64b8b333d6a976f0ea65c4949149d...
feature_engineering_complete,False
fingerprint,4dbb07439ab6c3c986bf321a5459650055b1aff61cc14f...


Installed-engine component checks: PASSED 736 fixtures


## 4 · Inspect activation and the cost of postponement

These views show coverage and action activation—not learned importance or independent observations. Multiple turns and both seats of one seed are correlated. The other days are descriptive coverage only; the live policy intervention is limited to days 20 and 21.

In [7]:
screen_plots = screen_figures(OUT)
figures.append(screen_plots[0])
screen_plots[0].show()

In [8]:
figures.append(screen_plots[1])
screen_plots[1].show()

In [9]:
figures.append(screen_plots[2])
screen_plots[2].show()

In [10]:
figures.append(screen_plots[3])
screen_plots[3].show()

In [11]:
figures.append(screen_plots[4])
screen_plots[4].show()

## 5 · One endpoint pair, only if the screen activates

Each arm replays 480 recorded prefix transitions in a fresh **official seeded environment**, verifying observations. It then advances 239 callbacks through the final state with responsive candidate and opponent policies. Official stochastic nights are handled by the real environment, not by copying future shops from a replay.

Maximum work: 960 prefix replays plus 478 responsive transitions across both arms. The control must reproduce the earlier source and notebook-10 final reward; the null fork must match the same-state control on every tested callback. Both arms retain the small final-sale correction; neither uses notebook 11's rejected route shortlist.

This pair uses already-examined development seeds. It can reject an idea or justify a next validation step, but cannot establish broad superiority.

In [12]:
pilot = run_stage("pilot")
display(pd.Series(pilot, name="live pilot result").to_frame())
print("DECISION:", pilot["decision"])
if pilot.get("completed_pairs"):
    display(pd.read_csv(OUT / "pilot/paired_results.csv"))
else:
    print("No new continuation: the activation screen did not qualify a source.")

{"utc": "2026-09-12T02:35:01.489185+00:00", "stage": "PREFLIGHT_PASSED", "verified_objects": 10}
{"utc": "2026-09-12T02:35:02.835498+00:00", "stage": "PREFIX_REPLAY", "mode": "control", "completed": 120, "total": 480}
{"utc": "2026-09-12T02:35:03.271549+00:00", "stage": "PREFIX_REPLAY", "mode": "control", "completed": 240, "total": 480}
{"utc": "2026-09-12T02:35:03.700422+00:00", "stage": "PREFIX_REPLAY", "mode": "control", "completed": 360, "total": 480}
{"utc": "2026-09-12T02:35:04.218579+00:00", "stage": "PREFIX_REPLAY", "mode": "control", "completed": 480, "total": 480}
{"utc": "2026-09-12T02:35:06.371590+00:00", "stage": "SUFFIX_PROGRESS", "mode": "control", "completed": 48, "total": 239}
{"utc": "2026-09-12T02:35:08.505362+00:00", "stage": "SUFFIX_PROGRESS", "mode": "control", "completed": 96, "total": 239}
{"utc": "2026-09-12T02:35:10.472025+00:00", "stage": "SUFFIX_PROGRESS", "mode": "control", "completed": 144, "total": 239}
{"utc": "2026-09-12T02:35:10.726569+00:00", "stage":

,live pilot result
candidate_callback_max_ms,21.472007
cloud_resources_modified,False
code_sha256,9c952471dc3d6a9185d5bda803955b323ca3c0cf0de426...
completed_pairs,1
decision,STOP_NEGATIVE_ENDPOINT
elapsed_seconds,26.886299
engine_sha256,bc8a54879ef02c7ea64b8b333d6a976f0ea65c4949149d...
feature_engineering_complete,False
fingerprint,4dbb07439ab6c3c986bf321a5459650055b1aff61cc14f...
full_seasons_recomputed_from_policy,0


DECISION: STOP_NEGATIVE_ENDPOINT


,seed,seat,opponent,arm,coins_control,coins_candidate,coins_delta,opponent_coins_control,opponent_coins_candidate,opponent_coins_delta,...,coin_margin_delta,local_match_score_control,local_match_score_candidate,local_match_score_delta,residual_product_units_control,residual_product_units_candidate,residual_product_units_delta,same_state_action_changes,water_commands_control,water_commands_candidate
0,1601,0,livestock_fertilizer,coordinated,44840.0,44444.0,-396.0,44082.0,43841.0,-241.0,...,-155.0,1.0,1.0,0.0,0,0,0,15,78,68


In [13]:
endpoint_plots = pilot_figures(OUT)
figures.append(endpoint_plots[0])
endpoint_plots[0].show()

In [14]:
figures.append(endpoint_plots[1])
endpoint_plots[1].show()

## 6 · Checkpoint, report, and stop

A negative result is preserved, not rerun until it passes. A positive result is provisional and must next face independent groups and distinct opponents. No official leaderboard score is computed here, and no features are declared final-selected.

The inherited callback restriction to at most three hired hands on either farm is still present. This experiment is not a broad submission-readiness certificate.

In [15]:
dashboard(figures, OUT / "maintenance_dashboard.html")
summary = {
    "screen_decision": screen["decision"], "pilot_decision": pilot["decision"],
    "completed_pairs": pilot.get("completed_pairs", 0),
    "official_submission_score": None, "github_updated": False,
    "feature_engineering_complete": False,
    "charts": len(figures), "dashboard": "outputs/maintenance_dashboard.html",
}
(OUT / "notebook_review.json").write_text(json.dumps(summary, indent=2))
print("NOTEBOOK12_COMPLETE")
display(pd.Series(summary, name="milestone decision").to_frame())

NOTEBOOK12_COMPLETE


,milestone decision
screen_decision,ELIGIBLE_FOR_ONE_PAIRED_PILOT
pilot_decision,STOP_NEGATIVE_ENDPOINT
completed_pairs,1
official_submission_score,None
github_updated,False
feature_engineering_complete,False
charts,8
dashboard,outputs/maintenance_dashboard.html


### Save and return the results

Press **Ctrl+S** now. Then use a JupyterLab terminal to run the command below; bundling also works if an earlier cell failed.

```bash
cd /home/sagemaker-user/kaggriculture_water_maintenance
PYTHON_BIN="$(python3 - <<'PY'
import json
from pathlib import Path
p = Path.home() / 'kaggriculture_manual_resume/state/runtime.json'
print(json.loads(p.read_text())['executable'])
PY
)"
"$PYTHON_BIN" run_maintenance.py bundle
```

Download **kaggriculture_water_maintenance_results.zip** (not the input ZIP). Then stop the existing app in **Studio → Running instances**. Do not delete the space. A notebook timeout or closed browser is not an AWS shutdown.

**References:** the pinned project `FeedPolicy`, `docs/feature_coverage.md`, and Kaggle's official interpreter. Source URLs, limitations, and the next research gates are recorded in `FEATURE_RESEARCH.md`.